Implement graph driver restore on reboot

crosbymichael opened the pull request:
This fixes reference counting for overlayfs, devicemapper, and zfs to restore the initial count by checking if the path that is to be reference counted is already mounted for when the daemon dies with running containers.  

It basically changes the reference counter from using the id of the layer to using the path so that we can check if it is reference counted or not.  

Overlay needed additional changes because it was using a path cache and that could be removed.

crosbymichael committed:
Restore ref count

Signed-off-by: Michael Crosby <crosbymichael@gmail.com>

crosbymichael committed:
Remove overlay pathCache

Signed-off-by: Michael Crosby <crosbymichael@gmail.com>

crosbymichael committed:
Remove restart test

This test is not applicable anymore now that containers are not stopped
when the daemon is restored.

Signed-off-by: Michael Crosby <crosbymichael@gmail.com>

cpuguy83 commented on daemon/graphdriver/counter.go:
```diff
@@ -1,32 +1,69 @@
+	// count if it is mounted as it is in use.
+	if !m.check {
+		m.check = true
+		mntd, _ := mount.Mounted(path)
```
This is super expensive.

crosbymichael replied:
sure, but there are many reasons why this is worth the cost for getting the correct results so i don't know what to tell ya

crosbymichael replied:
its only done once anyways per mount so i think it is fine

tonistiigi replied:
`graphdriver.Mounted` is faster and should be usable for overlay and zfs.

crosbymichael replied:
@tonistiigi ok, what about devmapper?

cpuguy83 commented:
There is already refcounting in the layer store, this seems like the best place to keep things since other subsystems should be interacting only with the layer store.
I'd added the ref counter at this level because of fsdiff, but it seems like we should be able to move this so the interactions from fsdiff go through the layer store.

crosbymichael commented:
@cpuguy83 you cannot do that at that level because these drivers mount themselves.

tonistiigi commented:
@crosbymichael This still requires #21732 to work properly, right? I wonder if we are ready to make the live-restoring default in master as there are still unresolved issues around networking.

crosbymichael commented:
@tonistiigi this works without that PR.  I verified using things like `diff` and the correct layer counts are updated without having to persist path data to disk and read it after restore.

crosbymichael committed:
Add fast path for fsmagic supported drivers

For things that we can check if they are mounted by using their fsmagic
we should use that and for others do it the slow way.

Signed-off-by: Michael Crosby <crosbymichael@gmail.com>

crosbymichael committed:
Add reference counting to aufs

Signed-off-by: Michael Crosby <crosbymichael@gmail.com>

crosbymichael commented:
@cpuguy83 @tonistiigi ok, i've added a fast path for checking if a path is mounted in the reference counter. 

I also added the reference counter to aufs as it is the last driver that does mounts but did not have any reference counting.  I just have to work on removing the code from the layer store because I don't think a solution like #21732 will be needed now as it wouldn't fit every driver anyways.

crosbymichael committed:
Remove ref counting from layer store

Signed-off-by: Michael Crosby <crosbymichael@gmail.com>

thaJeztah commented:
Looks like build is failing;

```
21:58:33 # github.com/docker/docker/layer
21:58:33 c:\go\src\github.com\docker\docker\layer\layer_test.go:679: rl.activityCount undefined (type *referencedRWLayer has no field or method activityCount)
21:58:33 c:\go\src\github.com\docker\docker\layer\layer_test.go:680: rl.activityCount undefined (type *referencedRWLayer has no field or method activityCount)
```

crosbymichael commented:
@jstarks @jhowardmsft 

for this error on tp5 tests:

```
23:46:31 ... "restart 56528aba7109c156db0432f1a9e67269ee9cde1cdd698874ba6e5f1da3622469" failed with errors: Error response from daemon: Cannot restart container 56528aba7109c156db0432f1a9e67269ee9cde1cdd698874ba6e5f1da3622469: hcsshim::ActivateLayer failed in Win32: The process cannot access the file because it is being used by another process. (0x20) id=56528aba7109c156db0432f1a9e67269ee9cde1cdd698874ba6e5f1da3622469 flavour=1
```

Is there a call that we can make to hcsshim that will tell us if the layer is already active or mounted?

jstarks commented:
@swernli Can you look into this?

crosbymichael committed:
Add windows graph driver ref counter

Signed-off-by: Michael Crosby <crosbymichael@gmail.com>

swernli commented:
LGTM for the Windows graphdriver.

thaJeztah commented:
ping @tonistiigi @cpuguy83 PTAL

tonistiigi commented:
cc @dmcgowan

dmcgowan commented on daemon/graphdriver/overlay/overlay.go:
```diff
@@ -339,77 +335,54 @@ func (d *Driver) Remove(id string) error {
 	if _, err := os.Stat(dir); err != nil {
 		return "", err
 	}
+	mergedDir := path.Join(dir, "merged")
```
Why place this before the "root" check. When there is a "root" there is no "merged" directory or any mount to track.

crosbymichael committed:
Fix overlay use of rootdir and defer

Check for the rootDir first because the mergeDir may not exist if root
is present.

Also fix unmounting in the defer to make sure it does not have a
refcount.

Signed-off-by: Michael Crosby <crosbymichael@gmail.com>

dmcgowan commented on layer/mounted_layer.go:
```diff
@@ -83,106 +82,30 @@ func (ml *mountedLayer) hasReferences() bool {
-	return nil
 }
 
 func (rl *referencedRWLayer) Mount(mountLabel string) (string, error) {
```
`Mount` and `Unmount` can just be removed

crosbymichael replied:
fixed

crosbymichael commented:
@dmcgowan any other comments/changes?  i'll make them in a batch when you are finished with the review

dmcgowan commented:
Just that one nit, otherwise overlay and layerstore LGTM

crosbymichael committed:
Remove mountedLayer Mount and Unmount

Signed-off-by: Michael Crosby <crosbymichael@gmail.com>

rhatdan commented:
Probably a good idea to ask @rhvgoyal for review of this.

crosbymichael commented:
@rhvgoyal can you please checkout the devmapper changes made in this pr?

tonistiigi commented on daemon/graphdriver/zfs/zfs.go:
```diff
@@ -105,7 +105,7 @@ func Init(base string, opt []string, uidMaps, gidMaps []idtools.IDMap) (graphdri
 		uidMaps:          uidMaps,
 		gidMaps:          gidMaps,
-		ctr:              graphdriver.NewRefCounter(),
+		ctr:              graphdriver.NewRefCounter(graphdriver.NewFsChecker(graphdriver.FsMagicZfs)),
```
Is `zfs` not allowed on top of `zfs`? I think it's even required for the driver that the parent is also from `zfs`.

crosbymichael replied:
so btrfs and zfs will not work with the fs magic hacks?

tonistiigi replied:
Btrfs doesn't mount so shouldn't need refcounter and mount detection.

crosbymichael replied:
ahh true

tonistiigi commented on layer/layer_store.go:
```diff
@@ -502,15 +502,9 @@ func (ls *layerStore) ReinitRWLayer(l RWLayer) error {
 	ls.mountL.Lock()
```
Do we need this method at all?

crosbymichael replied:
removed

rhvgoyal commented on daemon/graphdriver/counter.go:
```diff
@@ -1,32 +1,69 @@
-	count := c.counts[id]
+	m := c.counts[path]
+	if m == nil {
+		m = &minfo{check: true}
```
Should the initial state of check be "false" instead of "true"? IIUC, if a path is not 
in the c.counts map (m == nil), you want to check if it is mounted or not. Have I misunderstood it.

rhvgoyal commented on daemon/graphdriver/counter.go:
```diff
@@ -1,32 +1,69 @@
+	// count if it is mounted as it is in use.
+	if !m.check {
+		m.check = true
+		mntd, _ := mount.Mounted(path)
```
Curious that why do we need this mounted checks in Decrement() path. If daemon crashed and came back up, Decrement() should never be called before Increment(). If Decrement() is called before Increment(), this sounds like a bug in the code.

crosbymichael commented:
@rhvgoyal you are commenting on outdated commits

rhvgoyal commented:
@crosbymichael  Oops. Let me look at the latest one.

crosbymichael commented:
@rhvgoyal no problem, just wanted to let you know before you got too far

rhvgoyal commented on daemon/graphdriver/counter.go:
```diff
@@ -27,7 +27,7 @@ func (c *RefCounter) Increment(path string) int {
 	m := c.counts[path]
 	if m == nil {
-		m = &minfo{check: true}
+		m = &minfo{}
```
Ok, Found it. It has been fixed in this patch.

rhvgoyal replied:
Hmmm.., even this is outdated? What am I doing wrong. I pulled in your graph-restore branch locally and looking at it.

rhvgoyal commented on daemon/graphdriver/counter.go:
```diff
@@ -2,31 +2,66 @@ package graphdriver
-	count := c.counts[id]
+	m := c.counts[path]
+	if m == nil {
+		m = &minfo{}
```
I hope this is latest code I am commenting on.

Should we throw a warning here. Decrement() called before Increment(). Might help with solving some bugs. Can't think when Decrement() will be called before Increment() and it is not a bug.

crosbymichael replied:
In the case of a daemon restore this can happen.  Since the container is already running and mounted when it stops there is a call to Unmount but there never was a mount.

rhvgoyal commented on daemon/graphdriver/counter.go:
```diff
@@ -2,31 +2,66 @@ package graphdriver
+			m.count++
+		}
+	}
+	m.count--
```
Assume m==nil and path is mounted. In that case we will first increment m.count, then decrement it and still leave the path mounted. That can lead to problems in next Increment() operation which will think path is not mounted.

IOW, I am not sure how this mounted check is helping in case of Decrement().

crosbymichael replied:
I'm not sure how it would leave it mounted if the count goes: 

0, +1, -1, 0 == unmount

crosbymichael replied:
I guess I see what you are saying about having this in dec, its basically a noop

tonistiigi commented:
LGTM

cpuguy83 commented on libcontainerd/client_linux.go:
```diff
@@ -380,6 +380,81 @@ func (clnt *client) getOrCreateExitNotifier(containerID string) *exitNotifier {
+
+func (clnt *client) Restore(containerID string, options ...CreateOption) error {
+	cont, err := clnt.getContainerdContainer(containerID)
+	if err == nil && cont.Status != "stopped" {
```
Is this state defined as a const somewhere?

cpuguy83 replied:
I know it's not likely to ever change, but figure if we have the states defined might as well use it.

cpuguy83 replied:
Funny, all but "stopped" seems defined :)

https://github.com/crosbymichael/docker/blob/8bb4d31b10e4c3abee9ca92535461859bbf25d46/libcontainerd/types.go#L6

tonistiigi replied:
This is containerd state. https://github.com/docker/containerd/blob/126e5c403e863ed696e2074c530cb941afd27a96/runtime/runtime.go#L73

crosbymichael replied:
Also this is just moved code

cpuguy83 commented:
LGTM

LK4D4 commented:
It has merge label and passed CI. I'm merging.
